spec/validators/instrument_type_validator_spec.rb

```diff
@@ -0,0 +1,215 @@
+      it 'minimum or maximum number of wells' do
+        wells = [build(:pacbio_well, row: 'A', column: '1'), build(:pacbio_well, row: 'B', column: '1'), build(:pacbio_well, row: 'C', column: '1'), build(:pacbio_well, row: 'D', column: '1')]
+        run = build(:pacbio_run, system_name: 'Revio', plates: build_list(:pacbio_plate, 2, wells: [wells.first]))
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates]).to be_empty
+
+        run = build(:pacbio_run, system_name: 'Revio', plates: build_list(:pacbio_plate, 2, wells:))
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates]).to be_empty
+
+        # Minimum
+        run = build(:pacbio_run, system_name: 'Revio', plates: [build(:pacbio_plate, well_count: 0), build(:pacbio_plate, wells: [wells.first])])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].length).to eq(2)
+        expect(run.errors.messages[:plates]).to include("plate #{run.plates.first.plate_number} wells must have at least 1 well")
+
+        # Maximum
+        run = build(:pacbio_run, system_name: 'Revio', plates: [build(:pacbio_plate, well_count: 5), build(:pacbio_plate, wells:)])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].length).to eq(3)
+        expect(run.errors.messages[:plates]).to include("plate #{run.plates.first.plate_number} wells must have at most 4 wells")
+      end
+
+      it 'well positions' do
+        run = build(:pacbio_revio_run, well_positions_plate_1: ['G1'], well_positions_plate_2: ['A1'])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].length).to eq(2)
+        expect(run.errors.messages[:plates]).to include("plate #{run.plates.first.plate_number} wells must be in positions #{instrument_types['revio']['models']['wells']['validations']['well_positions']['options']['valid_positions'].join(',')}")
+
+        run = build(:pacbio_revio_run, system_name: 'Revio', well_positions_plate_1: %w[A1 B1 C1 H1], well_positions_plate_2: ['A1'])
+        instrument_type_validator = described_class.new(instrument_types:)
+        instrument_type_validator.validate(run)
+        expect(run.errors.messages[:plates].length).to eq(2)
+        expect(run.errors.messages[:plates]).to include("plate #{run.plates.first.plate_number} wells must be in positions #{instrument_types['revio']['models']['wells']['validations']['well_positions']['options']['valid_positions'].join(',')}")
+      end
+
+      context 'well combinations' do
```

Comment:
Can these similar tests, just with different combinations, be refactored a bit?

Reply:
Similar to [here](https://github.com/sanger/traction-service/pull/1019/files#diff-a7f21fa0f59d553c6c700dcc43f31291a776f2f23c3ebb3a58006039be6d5ae4R31)

Reply:
I have moved invalid_combinations into config and used those in the tests.